Move off of UIKit SPI: `-_pointerInteraction:regionForRequest:defaultRegion:completion:`

whsieh opened the pull request:
#### 5063252ff004392b50d68bfc2aa7d7d44564a5b0
<pre>
Move off of UIKit SPI: `-_pointerInteraction:regionForRequest:defaultRegion:completion:`
<a href="https://bugs.webkit.org/show_bug.cgi?id=260662">https://bugs.webkit.org/show_bug.cgi?id=260662</a>
rdar://114331132

Reviewed by Tim Horton.

Refactor some iOS-specific logic for computing pointer regions when using `UIPointerInteraction`,
such that we no longer require this UIKit SPI delegate method for asynchronously retrieving regions:
`-_pointerInteraction:regionForRequest:defaultRegion:completion:`. Rather than performing the hit-
test and then invoking the completion handler when the results arrive from the web process, we
instead cache the last computed `UIPointerRegion` on the content view and continuously perform
position information updates to update this cached `UIPointerRegion`, until we reach a state where
the most recent `UIPointerRegionRequest` is satisfied by the cached `_positionInformation`. This
actually allows us to simplify a bit of the logic here for updating pointer regions, since we&apos;ll now
simply return `_lastPointerRegion` from the delegate method, and continuously schedule position
information updates via `-doAfterPositionInformationUpdate:` until we catch up to the current
pointer location.

Test: iOSMouseSupport.BasicPointerInteractionRegions

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:

Remove SPI declarations that are now unused.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView cleanUpInteraction]):
(-[WKContentView pointerInteraction:regionForRequest:defaultRegion:]):
(-[WKContentView _updateLastPointerRegionIfNeeded:]):

Drive-by fix: use `[_lastPointerRegion rect]` instead of `self.bounds` when returning a region in
the case where the entire web view is `_editable`, but the position information is not yet up to
date. This shouldn&apos;t actually make any difference since the use of `&quot;WKEditablePointerRegion&quot;` means
that we&apos;ll always use `UIAxisNeither` to avoid nudging the cursor in any direction for this region,
but it makes visually debugging these pointer regions in Mail compose a bit easier, since we no
longer flicker rapidly between a pointer region the size of the web view that constraints neither
axis, and a pointer region the size of the expanded line rect that constrains to block direction,
when UIKit asks for `-pointerInteraction:styleForRegion:`.

(-[WKContentView pointerRegionForPositionInformation:point:]):
(-[WKContentView pointerInteraction:styleForRegion:]):

Another drive-by fix: it&apos;s actually possible in shipping iPadOS to end up in a situation where we
return an IBeam cursor that tries to align to the Y-axis, but the region is the entire bounds of the
view (which causes the cursor to disappear altogether until the user hovers over a different pointer
region). This is observable by hovering over the left edge of a `blockquote` on daringfireball.net.
After the changes in this patch (but without the tweak in this method), we no longer cause the
cursor to disappear, but we do get a visual flicker as the cursor attempts to animate to the center
of the view bounds. It&apos;s unclear whether this is better or worse, so it&apos;s probably better to just
fix this existing bug as a part of this refactoring:

The root cause is the fact that `-pointerRegionForPositionInformation:point:` can return a region
the size of `self.bounds` despite the position information containing an IBeam cursor type, if
`expandedLineRect.contains(location)` returns `false`, which is possible if the request location is
_just_ left or right of the `expandedLineRect`. To mitigate this, I&apos;m making it so that we don&apos;t
attempt to constrain to the region, if the region is taller (or wider, in vertical writing mode)
than an arbitrary length.

(-[WKContentView _pointerInteraction:regionForRequest:defaultRegion:completion:]): Deleted.
* Source/WebKit/UIProcess/ios/WKScrollView.mm:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/cursor-styles.html: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/iOSMouseSupport.mm:
(-[TestPointerRegionRequest initWithLocation:]):
(-[TestPointerRegionRequest location]):
(-[TestWKWebView pointerInteraction]):
(-[TestWKWebView pointerInfoAtLocation:]):

Additionally, write some test infrastructure and a basic API test to exercise this functionality by
calling directly into the `UIPointerInteractionDelegate` methods to grab the cursor styles and
regions for several common scenarios (e.g. CSS `cursor` property, editable regions, non-editable
text, links).

* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView elementRectFromSelector:]):
(-[WKWebView elementMidpointFromSelector:]):

Canonical link: <a href="https://commits.webkit.org/267361@main">https://commits.webkit.org/267361@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57f641a25a4748034c626f28b6c3a957b3240344

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16216 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16534 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16948 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17981 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15218 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16403 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19603 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16647 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17610 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16410 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16857 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13854 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18746 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14100 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14672 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21511 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15089 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14837 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/18072 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15437 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13090 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14665 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3925 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19029 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15260 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/iOSMouseSupport.mm:
```diff
@@ -134,6 +134,76 @@ - (void)_updateMouseTouches:(NSSet<UITouch *> *)touches;
+
+- (TestWebKitAPI::PointerInfo)pointerInfoAtLocation:(CGPoint)location
+{
+    NSLog(@"<WK>: getting pointer info at (%.0f, %.0f)", location.x, location.y);
```
I think we should remove this logging?

whsieh replied:
(Yep, just noticed this too 🤦 )

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/iOSMouseSupport.mm:
```diff
@@ -134,6 +134,75 @@ - (void)_updateMouseTouches:(NSSet<UITouch *> *)touches;
+    auto request = adoptNS([[TestPointerRegionRequest alloc] initWithLocation:location]);
+
+    RetainPtr defaultRegion = [UIPointerRegion regionWithRect:contentView.bounds identifier:nil];
+    [contentView pointerInteraction:interaction regionForRequest:request.get() defaultRegion:defaultRegion.get()];
+    [self waitForNextPresentationUpdate];
+
+    auto region = [contentView pointerInteraction:interaction regionForRequest:request.get() defaultRegion:defaultRegion.get()];
```
Hmm.. why are we calling `-[WKContentViewInteraction _pointerInteraction:]` twice? And throwing away the first result?

whsieh replied:
Ah, so the first call returns the last cached (stale) data, and asynchronously triggers a position information update; the second call (after ensuring a round trip to the web process and back) with the same request location should then yield the updated pointer region at that location.

aprotyas reviewed (commented):
We can remove this declaration from `UIKitSPI.h` now, right?

```objc
@protocol UIPointerInteractionDelegate_ForWebKitOnly <UIPointerInteractionDelegate>
@optional
- (void)_pointerInteraction:(UIPointerInteraction *)interaction regionForRequest:(UIPointerRegionRequest *)request defaultRegion:(UIPointerRegion *)defaultRegion completion:(void(^)(UIPointerRegion *region))completion;
@end
```

whsieh committed:
Move off of UIKit SPI: `-_pointerInteraction:regionForRequest:defaultRegion:completion:`
https://bugs.webkit.org/show_bug.cgi?id=260662
rdar://114331132

Reviewed by Tim Horton.

Refactor some iOS-specific logic for computing pointer regions when using `UIPointerInteraction`,
such that we no longer require this UIKit SPI delegate method for asynchronously retrieving regions:
`-_pointerInteraction:regionForRequest:defaultRegion:completion:`. Rather than performing the hit-
test and then invoking the completion handler when the results arrive from the web process, we
instead cache the last computed `UIPointerRegion` on the content view and continuously perform
position information updates to update this cached `UIPointerRegion`, until we reach a state where
the most recent `UIPointerRegionRequest` is satisfied by the cached `_positionInformation`. This
actually allows us to simplify a bit of the logic here for updating pointer regions, since we'll now
simply return `_lastPointerRegion` from the delegate method, and continuously schedule position
information updates via `-doAfterPositionInformationUpdate:` until we catch up to the current
pointer location.

Test: iOSMouseSupport.BasicPointerInteractionRegions

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:

Remove SPI declarations that are now unused.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView cleanUpInteraction]):
(-[WKContentView pointerInteraction:regionForRequest:defaultRegion:]):
(-[WKContentView _updateLastPointerRegionIfNeeded:]):

Drive-by fix: use `[_lastPointerRegion rect]` instead of `self.bounds` when returning a region in
the case where the entire web view is `_editable`, but the position information is not yet up to
date. This shouldn't actually make any difference since the use of `"WKEditablePointerRegion"` means
that we'll always use `UIAxisNeither` to avoid nudging the cursor in any direction for this region,
but it makes visually debugging these pointer regions in Mail compose a bit easier, since we no
longer flicker rapidly between a pointer region the size of the web view that constraints neither
axis, and a pointer region the size of the expanded line rect that constrains to block direction,
when UIKit asks for `-pointerInteraction:styleForRegion:`.

(-[WKContentView pointerRegionForPositionInformation:point:]):
(-[WKContentView pointerInteraction:styleForRegion:]):

Another drive-by fix: it's actually possible in shipping iPadOS to end up in a situation where we
return an IBeam cursor that tries to align to the Y-axis, but the region is the entire bounds of the
view (which causes the cursor to disappear altogether until the user hovers over a different pointer
region). This is observable by hovering over the left edge of a `blockquote` on daringfireball.net.
After the changes in this patch (but without the tweak in this method), we no longer cause the
cursor to disappear, but we do get a visual flicker as the cursor attempts to animate to the center
of the view bounds. It's unclear whether this is better or worse, so it's probably better to just
fix this existing bug as a part of this refactoring:

The root cause is the fact that `-pointerRegionForPositionInformation:point:` can return a region
the size of `self.bounds` despite the position information containing an IBeam cursor type, if
`expandedLineRect.contains(location)` returns `false`, which is possible if the request location is
_just_ left or right of the `expandedLineRect`. To mitigate this, I'm making it so that we don't
attempt to constrain to the region, if the region is taller (or wider, in vertical writing mode)
than an arbitrary length.

(-[WKContentView _pointerInteraction:regionForRequest:defaultRegion:completion:]): Deleted.
* Source/WebKit/UIProcess/ios/WKScrollView.mm:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/cursor-styles.html: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/iOSMouseSupport.mm:
(-[TestPointerRegionRequest initWithLocation:]):
(-[TestPointerRegionRequest location]):
(-[TestWKWebView pointerInteraction]):
(-[TestWKWebView pointerInfoAtLocation:]):

Additionally, write some test infrastructure and a basic API test to exercise this functionality by
calling directly into the `UIPointerInteractionDelegate` methods to grab the cursor styles and
regions for several common scenarios (e.g. CSS `cursor` property, editable regions, non-editable
text, links).

* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView elementRectFromSelector:]):
(-[WKWebView elementMidpointFromSelector:]):

Canonical link: https://commits.webkit.org/267361@main
